streamr/client/client.py

```diff
@@ -0,0 +1,416 @@
+
+    def __checkAutoDisconnect(self):
+        if self.options.get('autoDisconnect', False) is True and len(self.subsByStream.keys()) == 0:
+            logger.debug(
+                'Disconnecting due to no longer being subscribed to any streams')
+            self.disconnect()
+
+    def __resendAndSubscribe(self, sub):
+        if sub.getState() != Subscription.State.SUBSCRIBED and not sub.resending:
+
+            def subscribed():
+                if sub.hasResendOptions():
+                    self.__requestResend(sub)
+            sub.once('subscribed', subscribed)
+
+            self.__requestSubscribe(sub)
+
+    def __requestSubscribe(self, sub):
+
+        subs = self.subsByStream.get(sub.streamId, [])
+
+        subscribedSubs = []
+        for s in subs:
+            if s.getState() == Subscription.State.SUBSCRIBED:
+                subscribedSubs.append(s)
+
+        if len(subscribedSubs) == 0:
+            request = SubscribeRequest(
+                sub.streamId, 0, sub.apiKey, self.sessionToken)
+            logger.debug('_requestSubscribing client :%s' % (request))
+            self.connection.send(request)
+            sub.setState(Subscription.State.SUBSCRIBING)
+        elif len(subscribedSubs) > 0:
+            logger.debug(
+                '__requestSubscribe: another subscription for same stream : %s, insta-subscribing' % (sub.streamId))
+            sub.setState(Subscription.State.SUBSCRIBED)
+
+    def __requestUnsubscribe(self, streamId, partition=0, apiKey=None):
+        logger.debug('Client unsubscribing stream %s' % (streamId))
+        self.connection.send(UnsubscribeRequest(
+            streamId, partition, apiKey if apiKey is not None else self.options['apiKey'], self.sessionToken))
+
+    def __requestResend(self, sub, resendOptions=None):
+        sub.setResending(True)
+        request = ResendRequest(sub.streamId, sub.streamPartition, sub.id, resendOptions if resendOptions is not
+                                None else sub.getEffectiveResendOptions(), sub.apiKey, self.sessionToken)
+        logger.debug('__requestResend :%s' % (request))
+        self.connection.send(request)
+
+    def __requestPublish(self, streamId, data, apiKey):
+        request = PublishRequest(streamId, apiKey, self.sessionToken, data)
+        logger.debug('__requestPublish :%s' % (request))
+        self.connection.send(request)
+
+    def handleError(self, msg):
+        logger.debug(msg)
+        self.emit('error', msg)
+
+    def createStream(self, streamname, streamdes=None):
+        stream = creating(streamname, streamdes, self.sessionToken)
+        return stream
+
+    def getStreamByName(self, streamname):
+        return gettingByName(streamname, self.sessionToken)
+
+    def getStreamById(self, streamId):
+        return gettingById(streamId, self.sessionToken)
+
+    def getOrCreateStream(self, streamname):
+        stream = self.getStreamByName(streamname)
+        if len(stream) == 0 or stream is None:
```

Comment:
Only "stream is None" is needed, len(stream) will rise an error "TypeError: object of type 'NoneType' has no len()"

Reply:
We used len(stream) because a list type without any element is returned when using the getStreamByName method, rather than a None type. We revised the codes as the following.
```
    def getOrCreateStream(self, streamname):
        stream = self.getStreamByName(streamname)
        if stream is None or (isinstance(stream,list) and len(stream) == 0 ):
            return self.createStream(streamname)
        else:
            return stream
```